vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -42,28 +48,92 @@ type internal RoamingProfileStorageLocation(keyName: string) =
             let substituteLanguageName = if languageName = FSharpConstants.FSharpLanguageName then "FSharp" else languageName
             unsubstitutedKeyName.Replace("%LANGUAGE%", substituteLanguageName)
  
-[<Composition.Shared>]
-[<Microsoft.CodeAnalysis.Host.Mef.ExportWorkspaceServiceFactory(typeof<FSharpCheckerWorkspaceService>, Microsoft.CodeAnalysis.Host.Mef.ServiceLayer.Default)>]
-type internal FSharpCheckerWorkspaceServiceFactory
-    [<Composition.ImportingConstructor>]
+[<System.Composition.Shared>]
+[<ExportWorkspaceServiceFactory(typeof<IFSharpWorkspaceService>, ServiceLayer.Default)>]
+type internal FSharpWorkspaceServiceFactory
+    [<System.Composition.ImportingConstructor>]
     (
-        checkerProvider: FSharpCheckerProvider,
-        projectInfoManager: FSharpProjectOptionsManager
     ) =
-    interface Microsoft.CodeAnalysis.Host.Mef.IWorkspaceServiceFactory with
-        member _.CreateService(_workspaceServices) =
-            upcast { new FSharpCheckerWorkspaceService with
-                member _.Checker = checkerProvider.Checker
-                member _.FSharpProjectOptionsManager = projectInfoManager }
+
+    static let gate = obj()
```

Comment:
Please always document any locks very carefully.  Why is this `static`?  What resource is it protecting?

Reply:
Sure, that's fine. It's basically protecting to ensure we will never instantiate more than one FSharpChecker ever.